### DISCUSSION TITLE
Can't enable POST requests?

### DISCUSSION THREAD

**Discussion by pahu2353** [serwist/serwist#176] [Q&A]
Hey! Love serwist, it's been super helpful in getting my PWA started.

However, I can't seem to get POST requests working on my PWA... 

Here's my `sw.ts`:
```
import { defaultCache } from "@serwist/next/worker";
import type { PrecacheEntry, SerwistGlobalConfig } from "serwist";
import { Serwist } from "serwist";
import { NetworkFirst } from "serwist";

// This declares the value of `injectionPoint` to TypeScript.
// `injectionPoint` is the string that will be replaced by the
// actual precache manifest. By default, this string is set to
// `"self.__SW_MANIFEST"`.
declare global {
  interface WorkerGlobalScope extends SerwistGlobalConfig {
    __SW_MANIFEST: (PrecacheEntry | string)[] | undefined;
  }
}

declare const self: ServiceWorkerGlobalScope;

const serwist = new Serwist({
  precacheEntries: self.__SW_MANIFEST,
  skipWaiting: true,
  clientsClaim: true,
  navigationPreload: true,
  runtimeCaching: defaultCache,
});

// This should make all POST methods fetch from the server first. If the request fails, then it should cache responses with a 200 status code.
// Hopefully, this resolves the offline PWA issues. 
serwist.registerCapture(/^\/(api|auth|conference|setup)\/.*/, new NetworkFirst(), "POST");

serwist.addEventListeners();
```

When I go offline, the GET requests work perfectly. However, no POST requests work, despite me specifically "adding" them.

Looking at the cache in Dev Tool's Application, I can indeed see that the "POST" routes aren't being cached. However, I can also view the `sw.js` where it shows that this line is present: `ez.registerCapture(/^\/(api|auth|conference|setup)\/.*/, new ea, "POST"),`

I would really appreciate any help — I've been poring over this for the last week. Thank you so much in advance :)

**Comment by ZackTRobertson** [comment]


**Reply by DuCanhGH** [reply]
@pahu2353 appreciate your help, but you see, the solution you provided is... simply wrong 😅 Serwist partitions the routes by their `method`, which defaults to `"GET"` by the way, resulting in your route not matching any POST request at all. Use `method` instead. Perhaps this should be, well, better clarified in the docs. Also see below for an explanation :)

> Make sure that this entry comes before the defaultCache, as Serwist will match the first option it comes to and respond with that option.

That's true, but no `defaultCache` defines a route for POST requests :)

**Comment by DuCanhGH** [comment]
@pahu2353 you see, the problem is with the RegExp. When Serwist tries to match the route against the POST request, it calls the RegExp on the full URL rather than just the pathname, meaning that it always returns `false`. The easiest way to fix this is to remove the beginning "`^`" character, but you may want to use a `matcher` function instead:

```ts
const serwist = new Serwist({
  runtimeCaching: [
    {
      method: "POST",
      matcher({ url: { pathname }, sameOrigin }) {
        return sameOrigin && /^\/(api|auth|conference|setup)\/.*/.test(pathname);
      },
      handler: new NetworkFirst(),
    },
    ...defaultCache,
  ],
});
```

Hope this helps!

P/s: I'd rethink caching POST requests though. It's better to use [the Background Synchronization API](https://developer.mozilla.org/en-US/docs/Web/API/Background_Synchronization_API) so that POST requests are deferred until the user is online. [Here's how to use it with Serwist](https://serwist.pages.dev/docs/serwist/guide/background-syncing).

**Reply by ZackTRobertson** [reply]
Well, I "almost" had it 😅... Thanks for clarifying.

**Reply by pahu2353** [reply]
Thanks for the help! 

As you suggested, I ended up using the Background Synchronization API to defer POST requests until the user reconnects. Meanwhile, I did some optimistic rendering on my frontend to handle updates while the user is still offline. It ended up working pretty well 😄
